api/agent/agent.go

```diff
@@ -661,70 +647,91 @@ func callToHTTPRequest(ctx context.Context, call *call) *http.Request {
+		s.trySetError(err)
+		// first filter out timeouts
+		if ctx.Err() == context.DeadlineExceeded {
+			return context.DeadlineExceeded
 		}
-		common.Logger(ctx).WithField("resp", resp).Debug("Got resp from UDS socket")
+		return models.ErrFunctionResponse
+	}
+	defer resp.Body.Close()
 
-		// if ctx is canceled/timedout, then we close the body to unlock writeResp() below
-		defer resp.Body.Close()
+	common.Logger(ctx).WithField("resp", resp).Debug("Got resp from UDS socket")
 
-		ioErrChan := make(chan error, 1)
-		go func() {
-			ioErrChan <- writeResp(s.cfg.MaxResponseSize, resp, call.w)
-		}()
+	ioErrChan := make(chan error, 1)
+	go func() {
+		ioErrChan <- s.writeResp(ctx, s.cfg.MaxResponseSize, resp, call.respWriter)
+	}()
 
-		select {
-		case ioErr := <-ioErrChan:
-			errApp <- ioErr
-		case <-ctx.Done():
-			errApp <- ctx.Err()
+	select {
+	case ioErr := <-ioErrChan:
+		return ioErr
+	case <-ctx.Done():
+		if ctx.Err() == context.DeadlineExceeded {
+			// IMPORTANT: Container contract: If http-uds timeout, container cannot continue
+			s.trySetError(ctx.Err())
 		}
-	}()
-	return errApp
+		return ctx.Err()
+	}
 }
 
-func writeResp(max uint64, resp *http.Response, w io.Writer) error {
+func (s *hotSlot) writeResp(ctx context.Context, max uint64, resp *http.Response, w io.Writer) error {
 	rw, ok := w.(http.ResponseWriter)
 	if !ok {
+		// WARNING: this bypasses container contract translation. Assuming this is
+		// async mode, where we are storing response in call.stderr.
 		w = common.NewClampWriter(w, max, models.ErrFunctionResponseTooBig)
 		return resp.Write(w)
 	}
 
 	rw = newSizerRespWriter(max, rw)
 
+	// WARNING: is the following header copy safe?
 	// if we're writing directly to the response writer, we need to set headers
-	// and status code, and only copy the body. resp.Write would copy a full
+	// and only copy the body. resp.Write would copy a full
 	// http request into the response body (not what we want).
-
 	for k, vs := range resp.Header {
 		for _, v := range vs {
 			rw.Header().Add(k, v)
 		}
 	}
-	if resp.StatusCode > 0 {
-		rw.WriteHeader(resp.StatusCode)
+
+	var fdkErr error
+
+	// IMPORTANT: Container contract: Enforce 200/502/504 expections
+	switch resp.StatusCode {
+	case 200:
+		// FDK processed the request OK
+	case 502:
+		// FDK detected failure, container can continue
+		fdkErr = models.ErrFunctionFailed
+	case 504:
+		// FDK detected timeout, respond as if ctx expired, this gets translated & handled in handleCallEnd()
+		fdkErr = context.DeadlineExceeded
+	default:
+		// Any other code. Possible FDK failure. We shutdown the container
+		s.trySetError(fmt.Errorf("FDK Error, invalid status code %d", resp.StatusCode))
+		fdkErr = models.ErrFunctionInvalidResponse
 	}
-	_, err := io.Copy(rw, resp.Body)
-	return err
+
+	// despite errors above, we try to read the body, which could contain information why the request
+	// has failed, etc.
+	_, ioErr := io.Copy(rw, resp.Body)
```

Comment:
this is odd and maybe unnecessary, we should probably discard body for 504/502? (error handling here is odd to chase with ioErr and fdkErr)

Reply:
at least if we do this we buffer it into the thing at the moment - in the future this probably gets streamed back to client somewhere - should we define a response message for 502/504? at least it seems weird to write to the output and return an error from a contract level in dispatch

Reply:
also it'd make me more comfy (again, forgot about ioutil.Discard being done in resp.Body.Close) if we had a `defer io.Copy(ioutil.Discard, resp.Body)` above this thing in case we hit the clamp writer limit or something - at least, with go, have had fun times with conns being left open from forgetting to do this but i think recent versions they may have fixed so you might just ignore me now